src/zkevm_specs/state.py

```diff
@@ -263,6 +265,68 @@ def check_account_destructed(row: Row, row_prev: Row):
     # TODO: Missing constraints
 
 
+@is_circuit_code
+def check_tx_log(row: Row, row_prev: Row):
+    # tx_id | log_id | field_tag | index | value
+    tx_id = row.keys[1]
+    pre_tx_id = row_prev.keys[1]
+    log_id = row.keys[2]
+    pre_field_tag = row_prev.keys[3]
+    field_tag = row.keys[3]
+    index = row.keys[4]
+    pre_index = row_prev.keys[4]
+
+    # is_write is always true
+    assert row.is_write == 1
+    # reset log_id when tx_id increases
+    if row.tag() == row_prev.tag():
+        if tx_id != pre_tx_id:
+            assert tx_id == pre_tx_id + 1
+            assert log_id == 0
+        else:
+            # increase log_id when tag changes to Address, make sure tag can only increase(Non-Decreasing),
+            # when log_index stays same,
+            if pre_field_tag == U256(TxLogFieldTag.Address):
+                assert (field_tag - pre_field_tag).n > 0
+            # make sure if tag Data appear, data_index can only increase by one when tag stays same.
+            # make sure if tag Topic appear, topic_index in range [0,4),can only increase by one when tag stays same.
+            if field_tag == U256(TxLogFieldTag.Topic):
+                assert_in_range(index, 0, 3)
+            if field_tag not in [U256(TxLogFieldTag.Topic), U256(TxLogFieldTag.Data)]:
+                assert index == 0
+            elif pre_field_tag == field_tag:
+                assert index == pre_index + 1
+
+        # both log_index topic_index data_index are non negative, all starts with zero
+    assert index.n >= 0 and log_id.n >= 0
```

Comment:
yes, All values in the field are non-negative,  here can not use range check because `log_id ` and `index`(data index) are not exact clear upper bound,  I think better to remove this item checking for now ?

Reply:
> Ah, yes, All values in the field are non-negative, here can not use range check because `log_id ` and `index`(data index) are not exact clear upper bound, I think better to remove this item checking for now ?

Removing this constraint sounds good :)

Reply:
removed in [7208290](https://github.com/appliedzkp/zkevm-specs/pull/159/commits/72082901f7731cff2fe9eeb0f8ecc4833adb8f78)